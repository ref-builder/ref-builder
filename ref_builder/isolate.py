from enum import StrEnum


class IsolateNameType(StrEnum):
    """Possible types for isolate names.

    **Ordered by priority**. Do not reorder attributes.

    Isolate name types were previously called "source types". They are referred to this
    way in Virtool.
    """

    ISOLATE = "isolate"
    STRAIN = "strain"
    CLONE = "clone"
    VARIANT = "variant"
    GENOTYPE = "genotype"
    SEROTYPE = "serotype"
