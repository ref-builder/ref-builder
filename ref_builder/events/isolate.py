from pydantic import UUID4

from ref_builder.events.base import EventData, Event, IsolateQuery
from ref_builder.utils import IsolateName


class CreateIsolateData(EventData):
    """The data for a :class:`CreateIsolate` event."""

    id: UUID4
    legacy_id: str | None
    name: IsolateName | None


class CreateIsolate(Event):
    """An event that creates an isolate for a specific OTU."""

    data: CreateIsolateData
    query: IsolateQuery


class DeleteIsolateData(EventData):
    """The data for a :class:`DeleteIsolate` event."""

    rationale: str


class DeleteIsolate(Event):
    """An isolate deletion event."""

    data: DeleteIsolateData
    query: IsolateQuery
