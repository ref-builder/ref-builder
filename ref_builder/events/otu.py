from pydantic import UUID4

from ref_builder.events.base import EventData, Event, OTUQuery
from ref_builder.models import Molecule
from ref_builder.plan import MonopartitePlan, MultipartitePlan


class CreateOTUData(EventData):
    """The data for a :class:`CreateOTU` event."""

    id: UUID4
    acronym: str
    legacy_id: str | None
    molecule: Molecule
    name: str
    taxid: int
    plan: MonopartitePlan | MultipartitePlan


class CreateOTU(Event):
    """An event that creates a new OTU."""

    data: CreateOTUData
    query: OTUQuery


class ExcludeAccessionData(EventData):
    """The data for a :class:`ExcludeAccession` event."""

    accession: str


class ExcludeAccession(Event):
    """An accession exclusion event.

    This event is emitted when a Genbank accession is not going to be allowed in the
    reference.
    """

    data: ExcludeAccessionData
    query: OTUQuery


class CreatePlanData(EventData):
    """The data for a :class:`CreatePlan` event."""

    plan: MonopartitePlan | MultipartitePlan


class CreatePlan(Event):
    """An event that sets the isolate plan for an OTU."""

    data: CreatePlanData
    query: OTUQuery


class SetReprIsolateData(EventData):
    """The data for a :class:`SetReprIsolate` event."""

    isolate_id: UUID4


class SetReprIsolate(Event):
    """An event that sets the representative isolate for an OTU."""

    data: SetReprIsolateData
    query: OTUQuery
