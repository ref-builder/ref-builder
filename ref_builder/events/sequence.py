from pydantic import UUID4

from ref_builder.events.base import EventData, Event, SequenceQuery
from ref_builder.utils import Accession


class CreateSequenceData(EventData):
    """The data for a :class:`CreateSequence` event."""

    id: UUID4
    accession: Accession
    definition: str
    legacy_id: str | None
    segment: str
    sequence: str


class CreateSequence(Event):
    """An event that creates a sequence for a specific isolate and OTU."""

    data: CreateSequenceData
    query: SequenceQuery


class DeleteSequenceData(EventData):
    """The data for a :class:`DeleteSequence` event."""

    replacement: UUID4
    rationale: str


class DeleteSequence(Event):
    """An event that deletes a sequence.

    The second part of a sequence replacement.
    """

    data: DeleteSequenceData
    query: SequenceQuery
