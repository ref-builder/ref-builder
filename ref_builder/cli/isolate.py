import sys
from pathlib import Path
from uuid import UUID

import click

from ref_builder.cli.utils import pass_repo
from ref_builder.cli.validate import validate_no_duplicate_accessions
from ref_builder.options import ignore_cache_option, path_option
from ref_builder.otu.isolate import (
    add_and_name_isolate,
    add_genbank_isolate,
    add_unnamed_isolate,
)
from ref_builder.otu.modify import delete_isolate_from_otu
from ref_builder.otu.utils import RefSeqConflictError
from ref_builder.repo import Repo, locked_repo
from ref_builder.utils import IsolateName, IsolateNameType


@click.group(name="isolate")
@click.pass_context
@path_option
def isolate(ctx: click.Context, path: Path) -> None:
    """Manage isolates."""
    ctx.obj = ctx.with_resource(locked_repo(path))


@isolate.command(name="create")  # type: ignore
@click.argument(
    "accessions_",
    callback=validate_no_duplicate_accessions,
    metavar="ACCESSIONS",
    nargs=-1,
    type=str,
    required=True,
)
@click.option("--taxid", type=int, required=True)
@click.option(
    "--unnamed",
    is_flag=True,
    default=False,
    help="ignore isolate names in Genbank sources",
)
@click.option(
    "--name",
    type=(IsolateNameType, str),
    help='an overriding name for the isolate, e.g. "isolate ARWV1"',
)
@ignore_cache_option
@pass_repo
def isolate_create(
    repo: Repo,
    accessions_: list[str],
    ignore_cache: bool,
    name: tuple[IsolateNameType, str] | None,
    taxid: int,
    unnamed: bool,
) -> None:
    """Create a new isolate using the given accessions."""
    otu_ = repo.get_otu_by_taxid(taxid)

    if otu_ is None:
        click.echo(f"OTU {taxid} not found.", err=True)
        sys.exit(1)

    if unnamed:
        add_unnamed_isolate(
            repo,
            otu_,
            accessions_,
            ignore_cache=ignore_cache,
        )

        sys.exit(0)

    if name is not None:
        isolate_name_type, isolate_name_value = name
        isolate_name_ = IsolateName(type=isolate_name_type, value=isolate_name_value)

        try:
            add_and_name_isolate(
                repo,
                otu_,
                accessions_,
                ignore_cache=ignore_cache,
                isolate_name=isolate_name_,
            )
            sys.exit(0)

        except RefSeqConflictError as e:
            click.echo(
                f"{e.isolate_name} already exists, but RefSeq items may be "
                "promotable.",
            )
            sys.exit(1)

    try:
        add_genbank_isolate(
            repo,
            otu_,
            accessions_,
            ignore_cache=ignore_cache,
        )
    except RefSeqConflictError as e:
        click.echo(
            f"{e.isolate_name} already exists, but RefSeq items may be promotable,",
        )
        sys.exit(1)


@isolate.command(name="delete")  # type: ignore
@click.argument("IDENTIFIER", type=str)
@pass_repo
def isolate_delete(repo: Repo, identifier: str) -> None:
    """Delete an isolate with a UUID corresponding to IDENTIFIER.

    IDENTIFIER is an unique isolate ID (>8 characters)
    """
    isolate_id = None
    try:
        isolate_id = UUID(identifier)
    except ValueError:
        pass

    if isolate_id is None:
        try:
            isolate_id = repo.get_isolate_id_by_partial(identifier)

        except InvalidInputError:
            click.echo("Partial ID segment must be at least 8 characters long.", err=True)
            sys.exit(1)

        except PartialIDConflictError as e:
            click.echo(e, err=True)

    if isolate_id is None:
        click.echo("Isolate could not be found.", err=True)
        sys.exit(1)

    if (otu_id := repo.get_otu_id_by_isolate_id(isolate_id)) is None:
        click.echo(f"The containing OTU could not be found.", err=True)
        sys.exit(1)

    if (otu_ := repo.get_otu(otu_id)) is None:
        click.echo(f"OTU not found.", err=True)
        sys.exit(1)

    delete_isolate_from_otu(repo, otu_, isolate_id)

    click.echo("Isolate deleted.")
