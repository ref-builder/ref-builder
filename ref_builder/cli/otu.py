import sys
from pathlib import Path
from uuid import UUID

import click
from click import Context
import structlog

from ref_builder.cli.validate import validate_no_duplicate_accessions
from ref_builder.console import print_otu, print_otu_list
from ref_builder.options import ignore_cache_option, path_option
from ref_builder.otu.create import create_otu
from ref_builder.otu.isolate import (
    add_unnamed_isolate,
    add_and_name_isolate,
    add_genbank_isolate,
)
from ref_builder.otu.modify import (
    exclude_accessions_from_otu,
    set_representative_isolate,
    add_segments_to_plan,
    resize_monopartite_plan,
)
from ref_builder.otu.update import auto_update_otu, promote_otu_accessions
from ref_builder.otu.utils import RefSeqConflictError
from ref_builder.plan import SegmentRule, SegmentName
from ref_builder.repo import Repo
from ref_builder.utils import IsolateNameType, IsolateName

logger = structlog.get_logger()


@click.group(name="otu")
def otu() -> None:
    """Manage OTUs."""


@otu.command(name="create")
@click.argument("TAXID", type=int)
@click.argument(
    "accessions_",
    callback=validate_no_duplicate_accessions,
    metavar="ACCESSIONS",
    nargs=-1,
    type=str,
    required=True,
)
@click.option("--acronym", type=str, default="")
@ignore_cache_option
@path_option
def otu_create(
    accessions_: list[str],
    acronym: str,
    ignore_cache: bool,
    path: Path,
    taxid: int,
) -> None:
    """Create a new OTU.

    OTUs are created from a list of accessions and a taxonomy ID. The associated Genbank
    records are fetched and used to create the first isolate and a plan.
    """
    repo = Repo(path)

    if len(accessions_) != len(set(accessions_)):
        click.echo("Duplicate accessions were provided.", err=True)
        sys.exit(1)

    try:
        create_otu(
            repo,
            taxid,
            accessions_,
            acronym=acronym,
            ignore_cache=ignore_cache,
        )
    except ValueError as e:
        click.echo(e, err=True)
        sys.exit(1)


@otu.command(name="get")
@click.argument("IDENTIFIER", type=str)
@path_option
def otu_get(identifier: str, path: Path) -> None:
    """Get an OTU by its unique ID or taxonomy ID."""
    try:
        identifier = int(identifier)
        otu_ = Repo(path).get_otu_by_taxid(identifier)
    except ValueError:
        otu_ = Repo(path).get_otu(UUID(identifier))

    if otu_ is None:
        click.echo("OTU not found.", err=True)
        sys.exit(1)

    print_otu(otu_)


@otu.command(name="list")
@path_option
def otu_list(path: Path) -> None:
    """List all OTUs in the repository."""
    print_otu_list(Repo(path).iter_minimal_otus())


@otu.group(invoke_without_command=True)
@click.argument("TAXID", type=int)
@path_option
@click.pass_context
def update(ctx: Context, path: Path, taxid: int) -> None:
    """Update the specified OTU with new data."""
    repo = Repo(path)

    ctx.ensure_object(dict)
    ctx.obj = {
        "REPO": repo,
        "TAXID": taxid,
    }

    if not repo.get_otu_id_by_taxid(taxid):
        click.echo(f"OTU {taxid} not found.", err=True)
        sys.exit(1)

    if ctx.invoked_subcommand is None:
        click.echo(ctx.get_help())


@update.command(name="automatic")  # type: ignore
@ignore_cache_option
@click.pass_context
def otu_auto_update(ctx: Context, ignore_cache: bool) -> None:
    """Automatically update an OTU with the latest data from NCBI."""
    repo = ctx.obj["REPO"]
    taxid = ctx.obj["TAXID"]

    otu_ = repo.get_otu_by_taxid(taxid)

    if otu_ is None:
        click.echo(f"OTU not found for Taxonomy ID {taxid}.", err=True)
        click.echo(f'Run "virtool otu create {taxid} --autofill" instead.')
        sys.exit(1)

    auto_update_otu(repo, otu_, ignore_cache=ignore_cache)


@update.command(name="promote")  # type: ignore
@ignore_cache_option
@click.pass_context
def otu_promote_accessions(
    ctx: Context,
    ignore_cache: bool,
) -> None:
    """Promote all RefSeq accessions within this OTU."""
    repo = ctx.obj["REPO"]

    otu_ = repo.get_otu_by_taxid(ctx.obj["TAXID"])

    promote_otu_accessions(repo, otu_, ignore_cache)


@update.command(name="isolate")  # type: ignore
@click.argument(
    "accessions_",
    callback=validate_no_duplicate_accessions,
    metavar="ACCESSIONS",
    nargs=-1,
    type=str,
    required=True,
)
@click.option(
    "--unnamed",
    is_flag=True,
    default=False,
    help="ignore isolate names in Genbank sources",
)
@click.option(
    "--name",
    type=(IsolateNameType, str),
    help='an overriding name for the isolate, e.g. "isolate ARWV1"',
)
@ignore_cache_option
@click.pass_context
def isolate_create(
    ctx: Context,
    accessions_: list[str],
    ignore_cache: bool,
    name: tuple[IsolateNameType, str] | None,
    unnamed: bool,
) -> None:
    """Create a new isolate using the given accessions."""
    repo = ctx.obj["REPO"]
    taxid = ctx.obj["TAXID"]

    otu_ = repo.get_otu_by_taxid(taxid)
    if otu_ is None:
        click.echo(f"OTU {taxid} not found.", err=True)
        sys.exit(1)

    if unnamed:
        add_unnamed_isolate(
            repo,
            otu_,
            accessions_,
            ignore_cache=ignore_cache,
        )

    if name is not None:
        isolate_name_type, isolate_name_value = name
        isolate_name_ = IsolateName(type=isolate_name_type, value=isolate_name_value)

        try:
            add_and_name_isolate(
                repo,
                otu_,
                accessions_,
                ignore_cache=ignore_cache,
                isolate_name=isolate_name_,
            )
        except RefSeqConflictError as err:
            click.echo(
                f"{err.isolate_name} already exists, but RefSeq items may be "
                "promotable.",
            )
            sys.exit(1)

    try:
        add_genbank_isolate(
            repo,
            otu_,
            accessions_,
            ignore_cache=ignore_cache,
        )
    except RefSeqConflictError as err:
        click.echo(
            f"{err.isolate_name} already exists, but RefSeq items may be promotable,",
        )
        sys.exit(1)

    except ValueError as e:
        click.echo(e, err=True)
        sys.exit(1)


@update.command(name="exclude")  # type: ignore
@click.argument(
    "accessions_",
    metavar="ACCESSIONS",
    nargs=-1,
    type=str,
    required=True,
)
@click.pass_context
def accession_exclude(
    ctx: Context,
    accessions_: list[str],
) -> None:
    """Exclude the given accessions from this OTU.

    Any duplicate accessions will be ignored. Only one exclusion per unique accession
    will be made.
    """
    repo = ctx.obj["REPO"]
    taxid = ctx.obj["TAXID"]

    otu_ = repo.get_otu_by_taxid(taxid)

    if otu_ is None:
        click.echo(f"OTU {taxid} not found.", err=True)
        sys.exit(1)

    exclude_accessions_from_otu(repo, otu_, accessions_)


@update.command(name="default")  # type: ignore
@click.argument("ISOLATE_KEY", type=str)
@click.pass_context
def otu_set_representative_isolate(
    ctx: Context,
    isolate_key: str,
) -> None:
    """Update the OTU with a new representative isolate."""
    taxid = ctx.obj["TAXID"]

    repo = ctx.obj["REPO"]

    otu_ = repo.get_otu_by_taxid(taxid)

    if otu_ is None:
        click.echo(f"OTU not found for Taxonomy ID {taxid}.", err=True)
        sys.exit(1)

    try:
        isolate_id = UUID(isolate_key)
    except ValueError:
        parts = isolate_key.split(" ")
        try:
            isolate_name = IsolateName(IsolateNameType(parts[0].lower()), parts[1])
        except ValueError:
            click.echo(f'Error: "{isolate_key}" is not a valid isolate name.', err=True)
            sys.exit(1)

        isolate_id = otu_.get_isolate_id_by_name(isolate_name)

    if isolate_id is None:
        click.echo("Isolate could not be found in this OTU.", err=True)
        sys.exit(1)

    set_representative_isolate(repo, otu_, isolate_id)


@update.group()
@click.pass_context
def plan(ctx: Context) -> None:
    """Add to and replace isolate plans for this OTU."""


@plan.command(name="extend")
@click.argument(
    "accessions_",
    metavar="ACCESSIONS",
    nargs=-1,
    type=str,
    required=True,
)
@click.option(
    "--optional",
    is_flag=True,
    help="Set additional segments as fully optional",
)
@click.pass_context
@ignore_cache_option
def plan_extend_segment_list(
    ctx: Context,
    accessions_: list[str],
    optional: bool,
    ignore_cache: bool,
) -> None:
    """Add recommended or optional segments to the OTU plan."""
    repo = ctx.obj["REPO"]
    taxid = ctx.obj["TAXID"]

    otu_ = repo.get_otu_by_taxid(taxid)
    if otu_ is None:
        click.echo(f"OTU {taxid} not found.", err=True)
        sys.exit(1)

    try:
        add_segments_to_plan(
            repo,
            otu_,
            rule=SegmentRule.OPTIONAL if optional else SegmentRule.RECOMMENDED,
            accessions=accessions_,
            ignore_cache=ignore_cache,
        )
    except ValueError as e:
        click.echo(e, err=True)


@plan.command(name="multipartite")
@click.argument(
    "accessions_",
    metavar="ACCESSIONS",
    nargs=-1,
    type=str,
    required=True,
)
@click.option(
    "--name",
    type=(str, str),
    help='a name for the originalsegment, e.g. "RNA A"',
)
@click.option(
    "--optional",
    is_flag=True,
    help="Set additional segments as fully optional",
)
@click.pass_context
@ignore_cache_option
def plan_resize_monopartite(
    ctx: Context,
    name: tuple[str, str],
    accessions_: list[str],
    optional: bool,
    ignore_cache: bool,
) -> None:
    """Change a monopartite OTU plan to a multipartite plan."""
    repo = ctx.obj["REPO"]
    taxid = ctx.obj["TAXID"]

    prefix, key = name

    otu_ = repo.get_otu_by_taxid(taxid)
    if otu_ is None:
        click.echo(f"OTU {taxid} not found.", err=True)
        sys.exit(1)

    try:
        resize_monopartite_plan(
            repo,
            otu_,
            name=SegmentName(prefix=prefix, key=key),
            rule=SegmentRule.OPTIONAL if optional else SegmentRule.RECOMMENDED,
            accessions=accessions_,
            ignore_cache=ignore_cache,
        )
    except ValueError as e:
        click.echo(e, err=True)
