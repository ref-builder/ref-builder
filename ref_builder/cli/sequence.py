import sys
from pathlib import Path

import click

from ref_builder.cli.options import path_option
from ref_builder.cli.utils import get_otu_sequence_ids_from_identifier, pass_repo
from ref_builder.console import print_sequence, print_sequence_as_json
from ref_builder.repo import Repo, locked_repo


@click.group(name="sequence")
@click.pass_context
@path_option
def sequence(ctx: click.Context, path: Path) -> None:
    """Manage sequences."""
    ctx.obj = ctx.with_resource(locked_repo(path))

@sequence.command(name="get")
@click.argument("IDENTIFIER", type=str)
@click.option("--id-only", is_flag=True, help="Only output sequence ID.")
@click.option(
    "--json",
    "json_",
    is_flag=True,
    help="Output in JSON form.",
)
@pass_repo
def sequence_get(repo: Repo, identifier: str, id_only: bool, json_: bool) -> None:
    """Get an isolate with a UUID corresponding to IDENTIFIER.

    IDENTIFIER can be an accession number or a unique isolate ID (>8 characters)
    """
    otu_id, sequence_id = get_otu_sequence_ids_from_identifier(repo, identifier)

    otu_ = repo.get_otu(otu_id)

    sequence_ = otu_.get_sequence_by_id(sequence_id)

    if sequence_ is None:
        click.echo("Sequence could not be found.", err=True)
        sys.exit(1)

    if id_only:
        click.echo(sequence_.id)

        sys.exit(0)

    if json_:
        print_sequence_as_json(sequence_)

    else:
        print_sequence(sequence_)
