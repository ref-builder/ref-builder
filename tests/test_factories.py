"""Tests for data factories used in testing."""

import uuid
from uuid import uuid4

from syrupy import SnapshotAssertion

from ref_builder.models.accession import Accession
from ref_builder.ncbi.models import NCBISource
from ref_builder.otu.builders.sequence import SequenceBuilder
from ref_builder.otu.validators.isolate import Isolate
from ref_builder.otu.validators.otu import OTU
from ref_builder.otu.validators.sequence import Sequence
from tests.fixtures.factories import (
    IsolateFactory,
    NCBIGenbankFactory,
    NCBISourceFactory,
    OTUFactory,
    SequenceFactory,
)


def test_ncbi_source_factory(ncbi_source_factory: NCBISourceFactory):
    """Test that NCBISourceFactory creates valid mock NCBISource objects."""
    assert all(
        isinstance(dummy_source, NCBISource)
        for dummy_source in ncbi_source_factory.coverage()
    )


def test_ncbi_genbank_factory(
    ncbi_genbank_factory: NCBIGenbankFactory, snapshot: SnapshotAssertion
):
    """Test that NCBIGenbankFactory creates valid fake Genbank records."""
    records = list(ncbi_genbank_factory.coverage())

    assert [
        SequenceBuilder(
            id=uuid4(),
            accession=Accession.from_string(record.accession_version),
            definition=record.definition,
            segment=uuid.uuid4(),
            sequence=record.sequence,
        )
        for record in records
    ]

    assert [record.model_dump() for record in records] == snapshot


class TestNCBIGenbankFactoryBuildIsolate:
    """Tests for NCBIGenbankFactory.build_isolate()."""

    def test_sequential_accessions(self, ncbi_genbank_factory: NCBIGenbankFactory):
        """Test that build_isolate() generates sequential accessions."""
        records = ncbi_genbank_factory.build_isolate(segment_count=3, refseq=True)

        accessions = [record.accession for record in records]

        # Extract numeric parts
        numbers = [int(acc.split("_")[1]) for acc in accessions]

        # Verify they are sequential
        assert numbers == [numbers[0], numbers[0] + 1, numbers[0] + 2]

    def test_consistent_version(self, ncbi_genbank_factory: NCBIGenbankFactory):
        """Test that all records have the same version."""
        records = ncbi_genbank_factory.build_isolate(segment_count=3, version=2)

        versions = [record.accession_version.split(".")[1] for record in records]

        assert all(v == "2" for v in versions)

    def test_letter_segment_keys(self, ncbi_genbank_factory: NCBIGenbankFactory):
        """Test that letter segment keys are generated correctly."""
        records = ncbi_genbank_factory.build_isolate(
            segment_count=3,
            segment_key_style="letter",
            segment_prefix="RNA",
            segment_delimiter="-",
        )

        segment_names = [record.source.segment for record in records]

        assert segment_names == ["RNA-A", "RNA-B", "RNA-C"]

    def test_number_segment_keys(self, ncbi_genbank_factory: NCBIGenbankFactory):
        """Test that number segment keys are generated correctly."""
        records = ncbi_genbank_factory.build_isolate(
            segment_count=3,
            segment_key_style="number",
            segment_prefix="DNA",
            segment_delimiter=" ",
        )

        segment_names = [record.source.segment for record in records]

        assert segment_names == ["DNA 1", "DNA 2", "DNA 3"]

    def test_shared_metadata(self, ncbi_genbank_factory: NCBIGenbankFactory):
        """Test that all records share the same organism, taxid, etc."""
        records = ncbi_genbank_factory.build_isolate(segment_count=3)

        # Check organism is consistent
        organisms = [record.organism for record in records]
        assert len(set(organisms)) == 1

        # Check taxid is consistent
        taxids = [record.source.taxid for record in records]
        assert len(set(taxids)) == 1

        # Check mol_type is consistent
        moltypes = [record.moltype for record in records]
        assert len(set(moltypes)) == 1

        # Check host is consistent
        hosts = [record.source.host for record in records]
        assert len(set(hosts)) == 1

    def test_refseq_accessions(self, ncbi_genbank_factory: NCBIGenbankFactory):
        """Test that RefSeq accessions are generated when refseq=True."""
        records = ncbi_genbank_factory.build_isolate(segment_count=3, refseq=True)

        assert all(record.accession.startswith("NC_") for record in records)
        assert all(record.refseq for record in records)

    def test_genbank_accessions(self, ncbi_genbank_factory: NCBIGenbankFactory):
        """Test that GenBank accessions are generated when refseq=False."""
        records = ncbi_genbank_factory.build_isolate(segment_count=3, refseq=False)

        assert all(not record.accession.startswith("NC_") for record in records)
        assert all(not record.refseq for record in records)

    def test_segment_count_validation(self, ncbi_genbank_factory: NCBIGenbankFactory):
        """Test that ValueError is raised when segment_count > 26 for letter style."""
        import pytest

        with pytest.raises(ValueError, match="segment_count cannot exceed 26"):
            ncbi_genbank_factory.build_isolate(
                segment_count=27, segment_key_style="letter"
            )

    def test_number_style_large_count(self, ncbi_genbank_factory: NCBIGenbankFactory):
        """Test that number style supports large segment counts."""
        records = ncbi_genbank_factory.build_isolate(
            segment_count=30,
            segment_key_style="number",
            segment_delimiter=" ",
        )

        assert len(records) == 30

        # Verify keys go up to 30
        segment_keys = [record.source.segment.split()[-1] for record in records]
        assert segment_keys[-1] == "30"
        assert segment_keys[0] == "1"

    def test_auto_prefix_from_moltype(self, ncbi_genbank_factory: NCBIGenbankFactory):
        """Test that segment prefix is auto-detected from mol_type."""
        from ref_builder.ncbi.models import NCBISourceMolType

        # Test DNA
        base_source = NCBISourceFactory.build(mol_type=NCBISourceMolType.GENOMIC_DNA)
        records = ncbi_genbank_factory.build_isolate(
            segment_count=2,
            base_source=base_source,
            segment_delimiter="-",
        )

        assert all(record.source.segment.startswith("DNA-") for record in records)

        # Test RNA
        base_source = NCBISourceFactory.build(mol_type=NCBISourceMolType.GENOMIC_RNA)
        records = ncbi_genbank_factory.build_isolate(
            segment_count=2,
            base_source=base_source,
            segment_delimiter="-",
        )

        assert all(record.source.segment.startswith("RNA-") for record in records)

    def test_snapshot(
        self,
        ncbi_genbank_factory: NCBIGenbankFactory,
        snapshot: SnapshotAssertion,
    ):
        """Test that build_isolate() generates expected data structure."""
        records = ncbi_genbank_factory.build_isolate(
            segment_count=3,
            refseq=True,
            segment_prefix="RNA",
            segment_key_style="letter",
            segment_delimiter="-",
            version=1,
        )

        assert [record.model_dump() for record in records] == snapshot


def test_sequence_factory(sequence_factory: SequenceFactory):
    """Test that SequenceFactory creates valid mock sequence data."""
    assert all(
        Sequence.model_validate(sequence.model_dump())
        for sequence in sequence_factory.coverage()
    )


def test_isolate_factory():
    """Test that IsolateFactory creates valid mock isolate data."""
    assert all(
        Isolate.model_validate(isolate.model_dump())
        for isolate in IsolateFactory.coverage()
    )


def test_otu_factory(otu_factory: OTUFactory):
    """Test that OTUFactory creates valid mock OTU data."""
    mock_otus = otu_factory.coverage()
    assert all(OTU.model_validate(otu.model_dump()) for otu in mock_otus)
