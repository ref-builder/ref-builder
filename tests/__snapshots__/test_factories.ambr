# serializer version: 1
# name: test_ncbi_genbank_factory
  list([
    dict({
      'accession': 'NC_914177',
      'accession_version': 'NC_914177.3',
      'comment': 'IPFaXVHtDTxOFvWvFcyb',
      'definition': 'PjpinXIHpKzfqrPnnLYY',
      'moltype': 'DNA',
      'organism': 'Maize bunchy top virus',
      'refseq': True,
      'sequence': 'GAAAGAGGCAHGCACTCTCATDGTGTSGTTAGAGBCAGHGCTTACTAACTCTGRAATVGTAGGMTGGCVTGAAYAGCGTTTGCBGACAAAGGGTCGTCAATGCKCTGGACAAGATVTAATGAYTDYTTCACAAVTCCTAYDDAACHCGCTCTTATVTCCSTTTTGGTTCCCTATACAACTGCGACGAWAGVGTACKTGAYATRGRGGGATAGCTADGCCGCTTTACTCATAATGTTCTACCCTGGTCCTTCRKTCAACGAGGTGGTGGCGGBCATACGAGGTCAGAHGCTYBAGGCGTKDTGTATAYBACTATTGAATACCGATCTCTGGGHCCGGCTTAGAGCDGHACCTCTCATCTTGGCCTAATCGGCVCGTGVTACCTATTTGCGGGCGCCCTCCSGGGGCTTGGCAGCCACGTTCACWGTGCAAGTAGCTCGACGGTCTCAGGTADGGARCTTGGACAHTYGCTCAAGTGTGGTATCAAACGAAGGDCTGAGATCTATGCCAGTCYNAGGTTCKTGGAYAACATACGRAAGATAAACGGGCTCDCTAMCTCCCATTAGTCGGGGTHAKGGAAGCTVACTATGGACAGAATGTAGGGTTTCTTGBCACTVGGGCGGTATCATCAGCTTTTGCCAYTTAHCYMDGKKGACCNGGGTSCTCHCAAGCRATGAACTATCCAAGCAGATGGTGCGGTGCSGGGNTAGGCCTGGCAGCATGAATGGDCCCCCGWTCTTCCCHACNGCTTSGCGKGTCGTGCTCATTAATGCTNTCGGTGCGCDGAADTATCWTGGACNCCTWDACTCATTCGGGGATCTCRAGATRACTGNTCGCAGTCAVCKDTCTGGTTGTAACCBCCATTGTATCTCAGGCATGCGGTCTAATACGCGTACGTAATCAAGVTCCATATGGAAGATTAAATMCTGTCTWTGCGCGTGCCCCTGCCACAATCTTCCTATCCCGGGATGAYAGAGGTGAGTTACT',
      'source': dict({
        'clone': '',
        'focus': True,
        'host': 'prunus',
        'isolate': '44-Jersey',
        'macronuclear': False,
        'mol_type': 'unassigned DNA',
        'organism': 'Maize bunchy top virus',
        'proviral': False,
        'segment': 'DNA A8',
        'strain': '',
        'taxid': 496185,
        'transgenic': False,
      }),
      'strandedness': 'single',
      'topology': 'linear',
    }),
    dict({
      'accession': 'NC_914177',
      'accession_version': 'NC_914177.3',
      'comment': 'edByzrBiKuuLRxEtRtsm',
      'definition': 'KQREBNCODYsyKWGfXTgc',
      'moltype': 'DNA',
      'organism': 'Maize bunchy top virus',
      'refseq': True,
      'sequence': 'TGTCGYCTTGTGTCATACCCGGACCACWTTTHTTGGTCCSTTACTAATGCSCTCGKACATCGGGTGNCTTAACACCATAATACACCNATCCAAAGACGAGCGAACGGCNGCGGGCTMTTAGATMTTTCAWBTACCCGTACGAMACATTACCCMGAAGTGATACTCTGAGCATTCAGAAKGATCGCTYVAWCCGCGGGVCCCTAACCTTTCTCCGAAGCTGGTTAGGGAACGTTTCCGVTAACGGSCCAGCATCCTTTBAWGATGGCCNACGTTCACACATMGATAGACTKCKYTACCTGTTTTTGGTGGTTGAGRWTACCTAGACGGACCTTCAGGAVTACGGCCARTCHCGCGTCMCTKCACAGATGCAWATTCTSACACBCCTCMCCCGGCTCACCGCCSGCGRGCATGTGTCTAGGCTCACTVAGAAGYTCATTTGGAGTCGAGTTTACGGGGCTTMHGGSTTATAGTCGTGTGGTAANTTGCSACGCTGGCAACCHMTGACCATAGACTGAAATGCCCTCCGATADGATACCAATTGATGATTGBAAATAAAGAGGSGCGGGGTTGKGTGAMGCTACTGTTTDTGTTTBCTTTACGACAAGTTTTCAATCAATGCAABCGTMTGCGTCATHGGGAGTCTGTTCAHACTAACCGCGGGTCAGAGTTAGACAGATGATCCAGGGMAGCRGCTCCCMCTTGATGTNCTVYTTGGWTAATATCTGGCTTAAATGGAVACGAC',
      'source': dict({
        'clone': '',
        'focus': True,
        'host': 'prunus',
        'isolate': '44-Jersey',
        'macronuclear': False,
        'mol_type': 'unassigned DNA',
        'organism': 'Maize bunchy top virus',
        'proviral': False,
        'segment': 'DNA A8',
        'strain': '',
        'taxid': 496185,
        'transgenic': False,
      }),
      'strandedness': 'double',
      'topology': 'circular',
    }),
  ])
# ---
